### PR TITLE
BatchDot, Basic Batchnorm, Fix BiasAdd, Fix Conv2D, CodeCleanup

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
Summary of changes:
1. Implement BatchDot operator.
2. Implement native MXNet basic Batchnorm operator. (This will be revisited)
3. Fix shape issue with BiasAdd operator.
4. Fix padding issue with Conv2D operator.
5. Add support for dilated conv2D
6. Remove buggy implementations - conv1D, deconv, conv3D
6. Some code clean up, code comments, renames and utility functions.

@nswamy @jesterhazy @Roshrini @yuruofeifei

### CODE REVIEW

File: keras/backend/mxnet_backend.py [hunk 1]
```diff
@@ -7,15 +7,15 @@
 from collections import defaultdict
 from keras.utils import conv_utils
 
-from .common import floatx, epsilon, set_epsilon, set_floatx, set_image_data_format, image_data_format
-
+from .common import floatx, epsilon, set_image_data_format, image_data_format
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: use absolute imports

File: keras/backend/mxnet_backend.py [hunk 2]
```diff
@@ -481,7 +233,7 @@ def variable(value, dtype=None, name=None, constraint=None):
 
     # MXNet backend do not support scalars
     if isinstance(value, np.ndarray) and len(value.shape) == 0:
-        raise ValueError("MXNet backend: Do not support scalars. Provided value for variable - ", value)
+        raise ValueError("MXNet Backend: Do not support scalars. Provided value for variable - ", value)
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: consider "Scalars are not supported"

File: keras/backend/mxnet_backend.py [hunk 3]
```diff
@@ -1302,34 +1054,47 @@ def batch_dot(x, y, axes=None):
...
-    if extra:
-        ret = squeeze(ret, 2)
-
-    return ret
+        diff = 0
+    if ndim(x) == 2 and ndim(y) == 2:
+        # MXNet supports only 3D. Expand_dims to make it 3D. At the end squeeze it back.
+        x = expand_dims(x, axis=1)
+        y = expand_dims(y, axis=2)
+        diff = True
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: mixed type variable (was int)

File: keras/backend/mxnet_backend.py [hunk 4]
```diff
@@ -1302,34 +1054,47 @@ def batch_dot(x, y, axes=None):
...
+            out = KerasSymbol(mx.sym.batch_dot(lhs=x.symbol, rhs=y.symbol, transpose_a=True))
+    else:
+        if axes is not None:
+            trans_x = False if axes[0] == ndim(x) - 1 else True
+            trans_y = True if axes[1] == ndim(y) - 1 else False
+        else:
+            trans_x = False
+            trans_y = False
+        out = KerasSymbol(mx.sym.linalg_gemm2(x.symbol, y.symbol, transpose_a=trans_x, transpose_b=trans_y))
+    if diff:
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: mixed type

File: keras/backend/mxnet_backend.py [hunk 5]
```diff
@@ -3264,12 +3039,14 @@ def pool2d(x, pool_size, strides=(1, 1),
...
     if data_format is None:
         data_format = image_data_format()
-    if data_format not in {"channels_first", "channels_last"}:
-        raise ValueError("`data_format` should be either `channels_first` or `channels_last`.")
-    if pool_mode not in {"max", "avg"}:
-        raise ValueError("`pool_mode` should be either `max` or `avg`.")
-    if padding not in {"same", "valid"}:
-        raise ValueError("`padding` should be either `same` or `valid`.")
+    _validate_data_format(data_format)
+
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: extra blank line

File: keras/backend/mxnet_backend.py [hunk 6]
```diff
@@ -3264,12 +3039,14 @@ def pool2d(x, pool_size, strides=(1, 1),
...
+    _validate_pool_mode(pool_mode)
+    _validate_padding_mode(padding)
+
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: extra blank line

File: keras/backend/mxnet_backend.py [hunk 7]
```diff
@@ -3302,12 +3079,15 @@ def pool3d(x, pool_size, strides=(1, 1, 1), padding='valid',
...
     if data_format is None:
         data_format = image_data_format()
-    if data_format not in {"channels_first", "channels_last"}:
-        raise ValueError("`data_format` should be either `channels_first` or `channels_last`.")
-    if pool_mode not in {"max", "avg"}:
-        raise ValueError("`pool_mode` should be either `max` or `avg`.")
-    if padding not in {"same", "valid"}:
-        raise ValueError("`padding` should be either `same` or `valid`.")
+    _validate_data_format(data_format)
+
```
Comment (jesterhazy) [inline keras/backend/mxnet_backend.py]: extra blank line

File: keras/backend/mxnet_backend.py (hunk 1 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: Following Keras recommendation. They follow relative imports.

File: keras/backend/mxnet_backend.py (hunk 2 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: Done

File: keras/backend/mxnet_backend.py (hunk 3 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: Ah! thanks good catch.

File: keras/backend/mxnet_backend.py (hunk 5 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: Fixed

File: keras/backend/mxnet_backend.py (hunk 6 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: Fixed

File: keras/backend/mxnet_backend.py (hunk 7 above)
Comment (sandeep-krishnamurthy) [reply keras/backend/mxnet_backend.py]: Fixed

File: keras/backend/mxnet_backend.py (hunk 1 above)
Comment (jesterhazy) [reply keras/backend/mxnet_backend.py]: sigh, but ok
